scripts/assign_haplotypes.py

```diff
@@ -0,0 +1,227 @@
+
+    args = parser.parse_args()
+
+    substitutions = read_metadata(
+        args.substitutions,
+        id_columns=args.metadata_id_columns,
+    )
+
+    if args.haplotype_column_name in substitutions.columns:
+        print(
+            f"ERROR: The requested column name for haplotype annotations, '{args.haplotype_column_name}', already exists in the substitutions table.",
+            file=sys.stderr,
+        )
+        sys.exit(1)
+
+    haplotype_definitions = pd.read_csv(
+        args.haplotypes,
+        sep='\t' if args.haplotypes.endswith('.tsv') else ',',
+        comment='#',
+        na_filter=False,
+        **PANDAS_READ_CSV_OPTIONS,
+    )
+
+    haplotype_definition_by_name = {}
+    for haplotype_name, haplotype_definition in haplotype_definitions.groupby("clade", sort=False):
+        definition = {}
+        for record in haplotype_definition.to_dict(orient="records"):
+            if record["gene"] == "clade":
+                # When a haplotype has a clade definition, all substitutions in
+                # the definition will be relative to that clade.
+                if "clade" in definition:
+                    print(
+                        f"ERROR: The haplotype '{haplotype_name}' has multiple clades in its definition which is not possible.",
+                        file=sys.stderr,
+                    )
+                    sys.exit(1)
+
+                definition["clade"] = record["site"]
+            elif record["gene"] == "nuc":
+                # Nucleotide substitutions look like "A7G", so we only need to
+                # compare each haplotype defining substitution to the position
+                # and derived allele.
+                if "nuc" not in definition:
+                    definition["nuc"] = []
+
+                definition["nuc"].append(str(record["site"]) + record["alt"])
+            else:
+                # Amino acid substitutions look like "HA1:Q173P", so we need to
+                # compare each haplotype defining substitution's gene and
+                # position/derived allele.
+                if "aa" not in definition:
+                    definition["aa"] = []
+
+                definition["aa"].append(
+                    (
+                        record["gene"],
+                        str(record["site"]) + record["alt"],
+                    ),
+                )
+
+        haplotype_definition_by_name[haplotype_name] = definition
+
+    assign_haplotype_per_record = partial(
+        assign_haplotype,
+        haplotype_definitions=haplotype_definition_by_name,
+        clade_column=args.clade_column,
+        default_haplotype=args.default_haplotype,
+        use_clade_as_default_haplotype=args.use_clade_as_default_haplotype,
+    )
+
+    # Assign haplotypes to each row.
+    substitutions[args.haplotype_column_name] = substitutions.apply(
+        assign_haplotype_per_record,
+        axis=1,
+    )
```

Comment:
[question] Did you gravitate to pandas here (because of speed, ergonomics etc) or just continue to use it because `read_metadata` returns a dataframe?

Reply:
Both because I had this script in mind as a new Augur command where `read_metadata` is the standard interface and for the ergonomics of the `apply` method.